twister: run tests on device with twister on windows

congnguyenhuu opened the pull request:
Currently, support for Twister on Windows is only build.
The only thing that needs to be ported is that select()
cannot wait on file descriptors on Windows. Therefore
the serial monitor function needs to be reworked to
support both OSes.

congnguyenhuu committed:
twister: run tests on device with twister on windows

Currently, support for Twister on Windows is only build.
The only thing that needs to be ported is that select()
cannot wait on file descriptors on Windows. Therefore
the serial monitor function needs to be reworked to
support both OSes.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

hakehuang commented:
@congnguyenhuu you need fix the check error

hakehuang commented:
@congnguyenhuu do you have a full guide on run twister on windows platform?

hakehuang commented:
> On Windows, it can run with these options via command line eg. `--device-testing --device-serial-baud 115200 --device-serial COM1` . Just the hardware map can not be generated via `--generate-hardware-map`.

I use the powershell, but it does not work.  so there need to be a full guide do you follow the steps https://docs.zephyrproject.org/latest/develop/getting_started/index.html to build in windows?

congnguyenhuu commented:
> > On Windows, it can run with these options via command line eg. `--device-testing --device-serial-baud 115200 --device-serial COM1` . Just the hardware map can not be generated via `--generate-hardware-map`.
> 
> I use the powershell, but it does not work. so there need to be a full guide do you follow the steps https://docs.zephyrproject.org/latest/develop/getting_started/index.html to build in windows?

yeah, I will add a full guide for it.

hakehuang commented:
> yeah, I will add a full guide for it.

this is my windows steps for frdm_k64f

1. install the cmake, ninja, west, dts
2.  open windows command prompt with adminstrator 
3. run below
python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
```
python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
Renaming output directory to C:\github\upstream\i2s\zephyr\twister-out.7
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.1.0-2269-gda47474c48b3
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\testplan.json

Device testing on:

| Platform   | ID   | Serial device   |
|------------|------|-----------------|
| frdm_k64f  |      | COM7            |

INFO    - JOBS: 8
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete: [32m   1/   1[39m  100%  skipped: [39m   0[39m, failed: [39m   0[39m
INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 77.68 seconds
INFO    - In total 1 test cases were executed, 0 skipped on 1 out of total 473 platforms (0.21%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board     | ID   |   Counter |
|-----------|------|-----------|
| frdm_k64f |      |         1 |
INFO    - Saving reports...
INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\twister.json
INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister.xml...
INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister_report.xml...
INFO    - Run completed
```

congnguyenhuu commented:
> > yeah, I will add a full guide for it.
> 
> this is my windows steps for frdm_k64f
> 
> 1. install the cmake, ninja, west, dts
> 2.  open windows command prompt with adminstrator 
> 3. run below
> python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> ```
> python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> Renaming output directory to C:\github\upstream\i2s\zephyr\twister-out.7
> INFO    - Using Ninja..
> INFO    - Zephyr version: zephyr-v3.1.0-2269-gda47474c48b3
> INFO    - Using 'zephyr' toolchain.
> INFO    - Building initial testsuite list...
> INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\testplan.json
> 
> Device testing on:
> 
> | Platform   | ID   | Serial device   |
> |------------|------|-----------------|
> | frdm_k64f  |      | COM7            |
> 
> INFO    - JOBS: 8
> INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
> INFO    - Adding tasks to the queue...
> INFO    - Added initial list of jobs to queue
> INFO    - Total complete: [32m   1/   1[39m  100%  skipped: [39m   0[39m, failed: [39m   0[39m
> INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 77.68 seconds
> INFO    - In total 1 test cases were executed, 0 skipped on 1 out of total 473 platforms (0.21%)
> INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
> 
> Hardware distribution summary:
> 
> | Board     | ID   |   Counter |
> |-----------|------|-----------|
> | frdm_k64f |      |         1 |
> INFO    - Saving reports...
> INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\twister.json
> INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister.xml...
> INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister_report.xml...
> INFO    - Run completed
> ```

Right as your steps. We install dependencies follow community. But It can be run with both powershell and command prompt without admin.
![tmp_fff682dd-6801-47d6-97b1-36c66be8710a.jpg](https://user-images.githubusercontent.com/76719459/180426327-3912e25b-bb4c-483d-80b6-3bf19787802a.jpg)

hakehuang commented:
> > > yeah, I will add a full guide for it.
> > 
> > 
> > this is my windows steps for frdm_k64f
> > 
> > 1. install the cmake, ninja, west, dts
> > 2. open windows command prompt with adminstrator
> > 3. run below
> >    python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> > 
> > ```
> > python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> > Renaming output directory to C:\github\upstream\i2s\zephyr\twister-out.7
> > INFO    - Using Ninja..
> > INFO    - Zephyr version: zephyr-v3.1.0-2269-gda47474c48b3
> > INFO    - Using 'zephyr' toolchain.
> > INFO    - Building initial testsuite list...
> > INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\testplan.json
> > 
> > Device testing on:
> > 
> > | Platform   | ID   | Serial device   |
> > |------------|------|-----------------|
> > | frdm_k64f  |      | COM7            |
> > 
> > INFO    - JOBS: 8
> > INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
> > INFO    - Adding tasks to the queue...
> > INFO    - Added initial list of jobs to queue
> > INFO    - Total complete: �[32m   1/   1�[39m  100%  skipped: �[39m   0�[39m, failed: �[39m   0�[39m
> > INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 77.68 seconds
> > INFO    - In total 1 test cases were executed, 0 skipped on 1 out of total 473 platforms (0.21%)
> > INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
> > 
> > Hardware distribution summary:
> > 
> > | Board     | ID   |   Counter |
> > |-----------|------|-----------|
> > | frdm_k64f |      |         1 |
> > INFO    - Saving reports...
> > INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\twister.json
> > INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister.xml...
> > INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister_report.xml...
> > INFO    - Run completed
> > ```
> 
> Right as your steps. We install dependencies follow community. But It can be run with both powershell and command prompt without admin. ![tmp_fff682dd-6801-47d6-97b1-36c66be8710a.jpg](https://user-images.githubusercontent.com/76719459/180426327-3912e25b-bb4c-483d-80b6-3bf19787802a.jpg)

can you add update document in the C:\github\zephyrproject\zephyr\doc\develop\getting_started  for windows run?

gopiotr commented:
@hakehuang @congnguyenhuu 

I tested this PR both in `Command Prompt` and in `Powershell` with and without administrative rights and so far everything works correctly to me. 

I used following commands for tests:

```
python scripts\twister -p nrf52840dk_nrf52840 -T samples/hello_world -vv --inline-logs -j 8 --device-testing  --device-serial COM4
```

and

```
python scripts\twister -p nrf52840dk_nrf52840 -T tests\kernel\common -s tests\kernel\common\kernel.common -vv --inline-logs -j 8 --device-testing --hardware-map map.yml
```

In second case I used following map file (generated by hand):
```yaml
- connected: true
  id: 000683593489
  platform: nrf52840dk_nrf52840
  product: J-Link
  runner: nrfjprog
  serial: COM4
```

I wondering if it could be valuable to give Twister users possibility to generate hardware map on Windows OS. After analyzing I observed that the problem (in my case with `nrf52840dk_nrf52840` platform) is connected with lack of information about USB product string ID of connected hardware. I observe this missing information only on Windows OS - on Linux everything works okay. Quick workaround for this can be as follow:

```diff
diff --git a/scripts/pylib/twister/twisterlib/hardwaremap.py b/scripts/pylib/twister/twisterlib/hardwaremap.py
index ec980771f6..cf74ce08e9 100644
--- a/scripts/pylib/twister/twisterlib/hardwaremap.py
+++ b/scripts/pylib/twister/twisterlib/hardwaremap.py
@@ -276,6 +276,10 @@ class HardwareMap:
                 # assume endpoint 0 is the serial, skip all others
                 if d.manufacturer == 'Texas Instruments' and not d.location.endswith('0'):
                     continue
+
+                if d.product is None:
+                    d.product = "unknown"
+
                 s_dev = DUT(platform="unknown",
                                         id=d.serial_number,
                                         serial=persistent_map.get(d.device, d.device),
```

And then generated hardware map on Windows can look like this:
```yaml
- connected: true
  id: 000683593489
  platform: unknown
  product: unknown
  runner: unknown
  serial: COM4
```

On the one hand it could be still valuable because it help to generate skeleton of hardware map file, but on the other hand such long list of `unknown` parameters can be misleading for beginner user - this solution requires that user know how to fill correctly all unknowns.

congnguyenhuu committed:
twister: support generate hardware map on windows

On windows, some board manufacturers have name that is "Microsoft"
instead of real its name. Need add it to manufacturer list.
Product information is missing. we can assign it "unknown" and
then add missing informations  by hand.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
doc: develop: test: update twister guide

add run steps guide for windows

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu commented:
> > > > yeah, I will add a full guide for it.
> > > 
> > > 
> > > this is my windows steps for frdm_k64f
> > > 
> > > 1. install the cmake, ninja, west, dts
> > > 2. open windows command prompt with adminstrator
> > > 3. run below
> > >    python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> > > 
> > > ```
> > > python scripts\twister -p frdm_k64f -T samples/hello_world  --device-testing --device-serial COM7  -T samples/hello_world/ --west-flash="--board-id=621000000,--erase"
> > > Renaming output directory to C:\github\upstream\i2s\zephyr\twister-out.7
> > > INFO    - Using Ninja..
> > > INFO    - Zephyr version: zephyr-v3.1.0-2269-gda47474c48b3
> > > INFO    - Using 'zephyr' toolchain.
> > > INFO    - Building initial testsuite list...
> > > INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\testplan.json
> > > 
> > > Device testing on:
> > > 
> > > | Platform   | ID   | Serial device   |
> > > |------------|------|-----------------|
> > > | frdm_k64f  |      | COM7            |
> > > 
> > > INFO    - JOBS: 8
> > > INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
> > > INFO    - Adding tasks to the queue...
> > > INFO    - Added initial list of jobs to queue
> > > INFO    - Total complete: �[32m   1/   1�[39m  100%  skipped: �[39m   0�[39m, failed: �[39m   0�[39m
> > > INFO    - 1 of 1 test configurations passed (100.00%), 0 failed, 0 skipped with 0 warnings in 77.68 seconds
> > > INFO    - In total 1 test cases were executed, 0 skipped on 1 out of total 473 platforms (0.21%)
> > > INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
> > > 
> > > Hardware distribution summary:
> > > 
> > > | Board     | ID   |   Counter |
> > > |-----------|------|-----------|
> > > | frdm_k64f |      |         1 |
> > > INFO    - Saving reports...
> > > INFO    - Writing JSON report C:\github\upstream\i2s\zephyr\twister-out\twister.json
> > > INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister.xml...
> > > INFO    - Writing xunit report C:\github\upstream\i2s\zephyr\twister-out\twister_report.xml...
> > > INFO    - Run completed
> > > ```
> > 
> > 
> > Right as your steps. We install dependencies follow community. But It can be run with both powershell and command prompt without admin. ![tmp_fff682dd-6801-47d6-97b1-36c66be8710a.jpg](https://user-images.githubusercontent.com/76719459/180426327-3912e25b-bb4c-483d-80b6-3bf19787802a.jpg)
> 
> can you add update document in the C:\github\zephyrproject\zephyr\doc\develop\getting_started for windows run?

I updated  run steps twister on windows in twister guide
[Test Runner (Twister) — Zephyr Project Documentation.pdf](https://github.com/zephyrproject-rtos/zephyr/files/9238823/Test.Runner.Twister.Zephyr.Project.Documentation.pdf)

congnguyenhuu commented:
> @hakehuang @congnguyenhuu
> 
> I tested this PR both in `Command Prompt` and in `Powershell` with and without administrative rights and so far everything works correctly to me.
> 
> I used following commands for tests:
> 
> ```
> python scripts\twister -p nrf52840dk_nrf52840 -T samples/hello_world -vv --inline-logs -j 8 --device-testing  --device-serial COM4
> ```
> 
> and
> 
> ```
> python scripts\twister -p nrf52840dk_nrf52840 -T tests\kernel\common -s tests\kernel\common\kernel.common -vv --inline-logs -j 8 --device-testing --hardware-map map.yml
> ```
> 
> In second case I used following map file (generated by hand):
> 
> ```yaml
> - connected: true
>   id: 000683593489
>   platform: nrf52840dk_nrf52840
>   product: J-Link
>   runner: nrfjprog
>   serial: COM4
> ```
> 
> I'm wondering if it could be valuable to give Twister users possibility to generate hardware map on Windows OS. After analyzing I observed that the problem (in my case with `nrf52840dk_nrf52840` platform) is connected with lack of information about USB product string ID of connected hardware. I observe this missing information only on Windows OS - on Linux everything works okay. Quick workaround for this can be as follow:
> 
> ```diff
> diff --git a/scripts/pylib/twister/twisterlib/hardwaremap.py b/scripts/pylib/twister/twisterlib/hardwaremap.py
> index ec980771f6..cf74ce08e9 100644
> --- a/scripts/pylib/twister/twisterlib/hardwaremap.py
> +++ b/scripts/pylib/twister/twisterlib/hardwaremap.py
> @@ -276,6 +276,10 @@ class HardwareMap:
>                  # assume endpoint 0 is the serial, skip all others
>                  if d.manufacturer == 'Texas Instruments' and not d.location.endswith('0'):
>                      continue
> +
> +                if d.product is None:
> +                    d.product = "unknown"
> +
>                  s_dev = DUT(platform="unknown",
>                                          id=d.serial_number,
>                                          serial=persistent_map.get(d.device, d.device),
> ```
> 
> And then generated hardware map on Windows can look like this:
> 
> ```yaml
> - connected: true
>   id: 000683593489
>   platform: unknown
>   product: unknown
>   runner: unknown
>   serial: COM4
> ```
> 
> On the one hand it could be still valuable because it help to generate skeleton of hardware map file, but on the other hand such long list of `unknown` parameters can be misleading for beginner user - this solution requires that user know how to fill correctly all unknowns.

Thanks for suggestion. I did it and mention missing product in docs

hakehuang reviewed (dismissed):
LGTM

manuargue commented:
@congnguyenhuu pls rebase and fix conflicts

congnguyenhuu commented:
@nashif @hakehuang @gopiotr I rebased and fixed conflicts. Pls help me review again.

stephanosio commented on doc/develop/test/twister.rst:
```diff
@@ -548,46 +606,106 @@ Executing tests on multiple devices
+
+.. tabs::
+
+   .. group-tab:: Ubuntu
```
ditto

stephanosio commented on doc/develop/test/twister.rst:
```diff
@@ -596,9 +714,20 @@ will be updated. This way you can use one single master hardware map and update
-  ./scripts/twister --device-testing --hardware-map map.yml -T samples/hello_world/
+.. tabs::
+
+   .. group-tab:: Ubuntu
```
ditto

stephanosio commented on doc/develop/test/twister.rst:
```diff
@@ -635,13 +764,24 @@ hardware map::
+
+.. tabs::
+
+   .. group-tab:: Ubuntu
```
ditto

stephanosio reviewed (changes requested):
Looks ok overall. Just some comments about the documentation.

congnguyenhuu commented on doc/develop/test/twister.rst:
```diff
@@ -36,17 +36,39 @@ These also affect the outputs of ``--testcase-report`` and
-::
+.. tabs::
+
+   .. group-tab:: Ubuntu
```
> Since this is applicable to all Linux distros and not specifically Ubuntu, it would be better to just specify Linux here:
> 
> ```diff
> -   .. group-tab:: Ubuntu
> +   .. group-tab:: Linux
> ```

yep, I updated.
